## IMPORTING NECCESARY LIBRARIES

In [1]:
import pandas as pd
import requests
import json

### MEASURED MAGNITUDES

To interpret the data of air quality, obtained from the Ministry for the Ecological Transition, we first have download some explanatory tables.

In [2]:
magnitudes_esp = pd.read_excel('./data/data_aire_esp/estacionesymagnitudes_2018_tcm30-501408.xlsx',sheet_name="Magnitudes",  header = 0)

In [27]:
magnitudes_esp = magnitudes_esp.drop('GRUPO', axis= 'columns')

KeyError: "['GRUPO'] not found in axis"

In [28]:
magnitudes_esp.columns = ['Codigo', 'Magnitud', 'Abreviatura o fórmula', 'Unidad medida']

In [30]:
magnitudes_esp['Magnitud'] = magnitudes_esp['Magnitud'].str.lower()
magnitudes_esp['Magnitud'] = magnitudes_esp['Magnitud'].str.capitalize()

In [59]:
magnitudes_esp

Codigo                         Magnitud Abreviatura o fórmula  \
0        1                Dióxido de azufre                   SO2   
1        6              Monóxido de carbono                    CO   
2        7            Monóxido de nitrógeno                    NO   
3        8             Dióxido de nitrógeno                   NO2   
4        9  Partículas en suspensión <2,5µm                 PM2.5   
5       10   Partículas en suspensión <10µm                  PM10   
6       12      Óxidos de nitrógeno totales                   NOX   
7       14                            Ozono                    O3   
8       17                  Arsenico (pm10)                    As   
9       19                     Plomo (pm10)                    Pb   
10      27            Benzo(a)pireno (pm10)                   BAP   
11      28                    Cadmio (pm10)                    Cd   
12      30                          Benceno                  C6H6   
13      62                    Níquel (pm10)                    NI   

   Unidad medida  
0          µg/m3  
1          mg/m3  
2          µg/m3  
3          µg/m3  
4          µg/m3  
5          µg/m3  
6      µg/m3 NO2  
7          µg/m3  
8          ng/m3  
9          µg/m3  
10         ng/m3  
11         ng/m3  
12         µg/m3  
13         ng/m3

In [6]:
magnitudes_mad = pd.read_excel('./data/magnitudes_unidades_tecnicas.xlsx', header = 0)

Codigo                  Magnitud Abreviatura o fórmula  Unidad medida   \
0       1       Dióxido de Azufre                     SO2         μg/m 3    
1       6     Monóxido de Carbono                      CO          mg/m3    
2       7  Monóxido de Nitrógeno                      NO          μg/m3    
3       8   Dióxido de Nitrógeno                     NO2         μg/m 3    
4       9     Partículas < 2.5 μm                   PM2.5          μg/m3    

           Técnica de medida   
0  Fluorescencia ultravioleta   
1       Absorción infrarroja   
2         Quimioluminiscencia   
3                         Id.   
4                Microbalanza

In [34]:
magnitudes_mad.columns = ['Codigo', 'Magnitud', 'Abreviatura o fórmula', 'Unidad medida', 'Técnica de medida']

In [35]:
magnitudes_mad = magnitudes_mad.drop('Técnica de medida', axis= 'columns')

In [36]:
magnitudes_mad.head()

Codigo                  Magnitud Abreviatura o fórmula Unidad medida
0       1       Dióxido de Azufre                   SO2        μg/m 3 
1       6     Monóxido de Carbono                    CO         mg/m3 
2       7  Monóxido de Nitrógeno                    NO         μg/m3 
3       8   Dióxido de Nitrógeno                   NO2        μg/m 3 
4       9     Partículas < 2.5 μm                 PM2.5         μg/m3

In [154]:
magnitudes = [magnitudes_esp, magnitudes_mad]
mags = pd.concat(magnitudes)
mags = mags.drop_duplicates('Codigo')
mags

Codigo                               Magnitud Abreviatura o fórmula  \
0        1                      Dióxido de azufre                   SO2   
1        6                    Monóxido de carbono                    CO   
2        7                  Monóxido de nitrógeno                    NO   
3        8                   Dióxido de nitrógeno                   NO2   
4        9        Partículas en suspensión <2,5µm                 PM2.5   
5       10         Partículas en suspensión <10µm                  PM10   
6       12            Óxidos de nitrógeno totales                   NOX   
7       14                                  Ozono                    O3   
8       17                        Arsenico (pm10)                    As   
9       19                           Plomo (pm10)                    Pb   
10      27                  Benzo(a)pireno (pm10)                   BAP   
11      28                          Cadmio (pm10)                    Cd   
12      30                                Benceno                  C6H6   
13      62                          Níquel (pm10)                    NI   
8       20                               Tolueno                   TOL    
10      35                           Etilbenceno                   EBE    
11      37                            Metaxileno                   MXY    
12      38                            Paraxileno                   PXY    
13      39                            Ortoxileno                   OXY    
14      42        Hidrocarburos totales (hexano)                   TCH    
15      43                                Metano                   CH4    
16      44  Hidrocarburos no metánicos (hexano)                  NMHC    

   Unidad medida  
0          µg/m3  
1          mg/m3  
2          µg/m3  
3          µg/m3  
4          µg/m3  
5          µg/m3  
6      µg/m3 NO2  
7          µg/m3  
8          ng/m3  
9          µg/m3  
10         ng/m3  
11         ng/m3  
12         µg/m3  
13         ng/m3  
8         μg/m3   
10        μg/m3   
11        μg/m3   
12        μg/m3   
13        μg/m3   
14        mg/m3   
15        mg/m3   
16        mg/m3

## DATA SPAIN

This data comes from the Ministerio Para la Transición Ecológica.
There is a dataset for each year and chemical element.

#### 2015 DATA

In [123]:
As_15esp = pd.read_excel('./data/data_aire_esp/datos2015_tcm30-425989/As_DD_2015.xlsx')

In [124]:
BaP_15esp = pd.read_excel('./data/data_aire_esp/datos2015_tcm30-425989/B(a)P_DD_2015.xlsx')

In [125]:
Cd_15esp = pd.read_excel('./data/data_aire_esp/datos2015_tcm30-425989/Cd_DD_2015.xlsx')

In [126]:
Ni_15esp = pd.read_excel('./data/data_aire_esp/datos2015_tcm30-425989/Ni_DD_2015.xlsx')

In [127]:
Pb_15esp = pd.read_excel('./data/data_aire_esp/datos2015_tcm30-425989/Pb_DD_2015.xlsx')

In [128]:
Pm25_15esp = pd.read_excel('./data/data_aire_esp/datos2015_tcm30-425989/PM2.5_DD_2015.xlsx')

In [129]:
Pm10_15esp = pd.read_excel('./data/data_aire_esp/datos2015_tcm30-425989/PM10_DD_2015.xlsx')

#### 2016 DATA

In [131]:
As_16esp = pd.read_excel('./data/data_aire_esp/dd_2016_tcm30/As_DD_2016.xlsx')

In [132]:
BaP_16esp = pd.read_excel('./data/data_aire_esp/dd_2016_tcm30/B(a)P_DD_2016.xlsx')

In [133]:
Cd_16esp = pd.read_excel('./data/data_aire_esp/dd_2016_tcm30/Cd_DD_2016.xlsx')

In [134]:
Ni_16esp = pd.read_excel('./data/data_aire_esp/dd_2016_tcm30/Ni_DD_2016.xlsx')

In [135]:
Pb_16esp = pd.read_excel('./data/data_aire_esp/dd_2016_tcm30/Pb_DD_2016.xlsx')

In [136]:
PM10_16esp = pd.read_excel('./data/data_aire_esp/dd_2016_tcm30/PM10_DD_2016.xlsx')

In [137]:
PM25_16esp = pd.read_excel('./data/data_aire_esp/dd_2016_tcm30/PM25_DD_2016.xlsx')

#### 2017 DATA

In [140]:
As_15esp = pd.read_csv('./data/data_aire_esp/dd_2017_tcm30-481444/As_DD_2017.csv')

In [141]:
BaP_17esp = pd.read_csv('./data/data_aire_esp/dd_2017_tcm30-481444/B(a)P_DD_2017.csv')

In [142]:
Cd_17esp = pd.read_csv('./data/data_aire_esp/dd_2017_tcm30-481444/Cd_DD_2017.csv')

In [143]:
Ni_17esp = pd.read_csv('./data/data_aire_esp/dd_2017_tcm30-481444/Ni_DD_2017.csv')

In [144]:
Pb_17esp = pd.read_csv('./data/data_aire_esp/dd_2017_tcm30-481444/Pb_DD_2017.csv')

In [145]:
PM25_17esp = pd.read_csv('./data/data_aire_esp/dd_2017_tcm30-481444/PM2.5_DD_2017.csv')

In [146]:
PM10_17esp = pd.read_csv('./data/data_aire_esp/dd_2017_tcm30-481444/PM10_DD_2017.csv')

#### 2018 DATA

In [55]:
As_18esp = pd.read_csv('./data/data_aire_esp/dd_2018_tcm30-501405/As_DD_2018.csv', sep = ';')

In [58]:
BAP_18esp = pd.read_csv('./data/data_aire_esp/dd_2018_tcm30-501405/B(a)P_DD_2018.csv', sep = ';')

In [60]:
Cd_18esp = pd.read_csv('./data/data_aire_esp/dd_2018_tcm30-501405/Cd_DD_2018.csv', sep = ';')

In [65]:
Ni_18esp = pd.read_csv('./data/data_aire_esp/dd_2018_tcm30-501405/Ni_DD_2018.csv', sep = ';')

In [66]:
Pb_18esp = pd.read_csv('./data/data_aire_esp/dd_2018_tcm30-501405/Pb_DD_2018.csv', sep = ';')

In [67]:
PM25_18esp = pd.read_csv('./data/data_aire_esp/dd_2018_tcm30-501405/PM2.5_DD_2018.csv', sep = ';')

In [68]:
PM10_18esp = pd.read_csv('./data/data_aire_esp/dd_2018_tcm30-501405/PM10_DD_2018.csv', sep = ';')

#### 2019 DATA

In [155]:
As_19esp = pd.read_csv('./data/data_aire_esp/diarios_2019_tcm30-513558/As_DD_2019.csv', sep = ';')

In [156]:
Bap_19esp = pd.read_csv('./data/data_aire_esp/diarios_2019_tcm30-513558/Bap_DD_2019.csv', sep = ';')

In [157]:
C6H6_19esp = pd.read_csv('./data/data_aire_esp/diarios_2019_tcm30-513558/C6H6_DD_2019.csv', sep = ';')

In [158]:
Cd_19esp = pd.read_csv('./data/data_aire_esp/diarios_2019_tcm30-513558/Cd_DD_2019.csv', sep = ';')

In [159]:
Ni_19esp = pd.read_csv('./data/data_aire_esp/diarios_2019_tcm30-513558/Ni_DD_2019.csv', sep = ';')

In [160]:
Pb_19esp = pd.read_csv('./data/data_aire_esp/diarios_2019_tcm30-513558/Pb_DD_2019.csv', sep = ';')

In [161]:
PM10_19esp = pd.read_csv('./data/data_aire_esp/diarios_2019_tcm30-513558/PM10_DD_2019.csv', sep = ';')

In [162]:
PM25_19esp = pd.read_csv('./data/data_aire_esp/diarios_2019_tcm30-513558/PM25_DD_2019.csv', sep = ';')

## OBTAINING DATA WITH API OF QUALITY OF AIR

In [165]:
response = requests.get("https://datos.madrid.es/egob/catalogo/keyword/aire.json?_page=0")
print(response.status_code)

200


In [166]:
print(response.json())

{'format': 'linked-data-api', 'version': '0.2', 'result': {'first': 'https://datos.madrid.es/egob/catalogo/keyword/aire.json?_page=0', 'items': [{'_about': 'https://datos.madrid.es/egob/catalogo/201410-0-calidad-aire-diario', 'description': [{'value': 'El Sistema Integral de la Calidad del Aire del Ayuntamiento de Madrid permite conocer en cada momento los niveles de contaminación atmosférica en el municipio. En este conjunto de datos puede obtener la información recogida por las estaciones de control de calidad del aire, con los datos diarios por anualidades de 2001 a 2021. (En el año en curso la información se actualizará mensualmente).   Los valores diarios publicados en el portal datos abiertos se calculan como media aritmética de los 24 valores horarios de cada día.   La legislación en materia de calidad del aire establece, para cada uno de los contaminantes regulados, distintos valore límite u objetivo expresados como medias horaria, diaria, octohoraria y anual.   En este portal 

In [167]:
urlcomp = 'https://datos.madrid.es/egob/catalogo/201410-10306612-calidad-aire-diario.csv'
headers = {'authorization': "Basic API Key Ommitted", 'accept': "application/json", 'accept': "text/csv"}

In [181]:
rcomp = requests.get(urlcomp, headers=headers)
data_21 = rcomp.text

In [183]:
AirQ_21_Mad = pd.DataFrame([x.split(';') for x in data_21.split('\n')])
AirQ_21_Mad.head()

0          1         2         3               4     5    6      7   \
0  PROVINCIA  MUNICIPIO  ESTACION  MAGNITUD  PUNTO_MUESTREO   ANO  MES    D01   
1         28        079         4         1   28079004_1_38  2021   01  00005   
2         28        079         4         1   28079004_1_38  2021   02  00010   
3         28        079         4         1   28079004_1_38  2021   03  00010   
4         28        079         4         1   28079004_1_38  2021   04  00009   

    8      9   ...     59   60     61   62     63   64     65   66     67   68  
0  V01    D02  ...    D27  V27    D28  V28    D29  V29    D30  V30    D31  V31  
1    V  00005  ...  00003    N  00000    N  00000    N  00000    N  00000    N  
2    N  00011  ...  00009    V  00009    V  00000    N  00000    N  00000    N  
3    V  00010  ...  00009    V  00009    V  00010    V  00010    V  00010    V  
4    V  00010  ...  00012    V  00012    V  00011    V  00011    V  00000    N  

[5 rows x 69 columns]

In [189]:
new_header = AirQ_21_Mad.iloc[0]
AirQ_21_Mad = AirQ_21_Mad[1:]
AirQ_21_Mad.columns = new_header
AirQ_21_Mad.head()

0 PROVINCIA MUNICIPIO ESTACION MAGNITUD PUNTO_MUESTREO   ANO MES    D01 V01  \
1        28       079        4        1  28079004_1_38  2021  01  00005   V   
2        28       079        4        1  28079004_1_38  2021  02  00010   N   
3        28       079        4        1  28079004_1_38  2021  03  00010   V   
4        28       079        4        1  28079004_1_38  2021  04  00009   V   
5        28       079        4        1  28079004_1_38  2021  05  00011   V   

0    D02  ...    D27 V27    D28 V28    D29 V29    D30 V30    D31 V31  
1  00005  ...  00003   N  00000   N  00000   N  00000   N  00000   N  
2  00011  ...  00009   V  00009   V  00000   N  00000   N  00000   N  
3  00010  ...  00009   V  00009   V  00010   V  00010   V  00010   V  
4  00010  ...  00012   V  00012   V  00011   V  00011   V  00000   N  
5  00011  ...  00010   V  00009   V  00009   V  00009   V  00009   V  

[5 rows x 69 columns]

In [174]:
AirQ_21_Mad.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 827 entries, 1 to 827
Data columns (total 69 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   PROVINCIA       827 non-null    object
 1   MUNICIPIO       826 non-null    object
 2   ESTACION        826 non-null    object
 3   MAGNITUD        826 non-null    object
 4   PUNTO_MUESTREO  826 non-null    object
 5   ANO             826 non-null    object
 6   MES             826 non-null    object
 7   D01             826 non-null    object
 8   V01             826 non-null    object
 9   D02             826 non-null    object
 10  V02             826 non-null    object
 11  D03             826 non-null    object
 12  V03             826 non-null    object
 13  D04             826 non-null    object
 14  V04             826 non-null    object
 15  D05             826 non-null    object
 16  V05             826 non-null    object
 17  D06             826 non-null    object
 18  V06       

In [184]:
urlcomp = 'https://datos.madrid.es/egob/catalogo/201410-10306609-calidad-aire-diario.csv'
headers = {'authorization': "Basic API Key Ommitted", 'accept': "application/json", 'accept': "text/csv"}

In [185]:
rcomp = requests.get(urlcomp, headers=headers)
data_20 = rcomp.text

In [191]:
AirQ_20_Mad = pd.DataFrame([x.split(';') for x in data_20.split('\n')])
AirQ_20_Mad.head()

0          1         2         3               4     5    6      7   \
0  PROVINCIA  MUNICIPIO  ESTACION  MAGNITUD  PUNTO_MUESTREO   ANO  MES    D01   
1         28        079         4         1   28079004_1_38  2020   01  00009   
2         28        079         4         1   28079004_1_38  2020   02  00008   
3         28        079         4         1   28079004_1_38  2020   03  00000   
4         28        079         4         1   28079004_1_38  2020   04  00000   

    8      9   ...     59   60     61   62     63   64     65   66     67   68  
0  V01    D02  ...    D27  V27    D28  V28    D29  V29    D30  V30    D31  V31  
1    V  00010  ...  00000    N  00000    N  00000    N  00006    N  00007    V  
2    V  00009  ...  00009    V  00011    V  00001    N  00000    N  00000    N  
3    N  00003  ...  00000    N  00000    N  00000    N  00000    N  00000    N  
4    N  00000  ...  00000    N  00000    N  00014    N  00008    V  00000    N  

[5 rows x 69 columns]

In [192]:
new_header = AirQ_20_Mad.iloc[0]
AirQ_20_Mad = AirQ_20_Mad[1:]
AirQ_20_Mad.columns = new_header
AirQ_20_Mad.head()

0 PROVINCIA MUNICIPIO ESTACION MAGNITUD PUNTO_MUESTREO   ANO MES    D01 V01  \
1        28       079        4        1  28079004_1_38  2020  01  00009   V   
2        28       079        4        1  28079004_1_38  2020  02  00008   V   
3        28       079        4        1  28079004_1_38  2020  03  00000   N   
4        28       079        4        1  28079004_1_38  2020  04  00000   N   
5        28       079        4        1  28079004_1_38  2020  05  00005   V   

0    D02  ...    D27 V27    D28 V28    D29 V29    D30 V30    D31 V31  
1  00010  ...  00000   N  00000   N  00000   N  00006   N  00007   V  
2  00009  ...  00009   V  00011   V  00001   N  00000   N  00000   N  
3  00003  ...  00000   N  00000   N  00000   N  00000   N  00000   N  
4  00000  ...  00000   N  00000   N  00014   N  00008   V  00000   N  
5  00004  ...  00004   V  00004   V  00004   V  00004   V  00004   V  

[5 rows x 69 columns]

## DATA MADRID

## DATA ANALISIS AND DATA CLEANING

In [ ]:
p_19 = pd.read_csv('./data/Indicadores_2019.csv', sep = ';')

In [ ]:
p_20 = pd.read_csv('./data/Indicadores_2020.csv', sep = ';')

In [ ]:
p_21 = pd.read_csv('./data/Indicadores_2021.csv', sep = ';')

In [ ]:
p_19.head()

In [ ]:
p_20.head()

In [ ]:
p_21.head()

In [ ]:
p_19.columns

In [ ]:
p_19.MAGNITUD.unique()

In [ ]:
p_19.PROVINCIA.unique()

In [ ]:
p_19.ESTACION.unique()

In [ ]:
m_t_e_19 = pd.read_excel('./data/matriculaciones_2019_anuario.xlsx', sheet_name = 'M_TUR_MEDIO', header = 2)
m_t_e_19.head()

In [ ]:
m_t_e_18 = pd.read_excel('./data/matriculaciones_2018_anuario.xlsx', sheet_name = 'M_TUR_MEDIO', header = 2)
m_t_e_18.head()

In [ ]:
m_t_e_17 = pd.read_excel('./data/matriculaciones_2017_anuario.xlsx', sheet_name = 'M_TUR_MEDIO', header = 2)
m_t_e_17.head()

In [ ]:
m_m_e_19 = pd.read_excel('./data/matriculaciones_2019_anuario.xlsx', sheet_name = 'M_MOTO_MEDIO', header = 2)
m_m_e_19.head()

In [ ]:
m_m_e_18 = pd.read_excel('./data/matriculaciones_2018_anuario.xlsx', sheet_name = 'M_MOTO_MEDIO', header = 2)
m_m_e_18.head()

In [ ]:
m_m_e_17 = pd.read_excel('./data/matriculaciones_2017_anuario.xlsx', sheet_name = 'M_MOTO_MEDIO', header = 2)
m_m_e_17.head()

### Explicación Magnitudes:
Abreviatura o fórmula
Unidad medida
Técnica de medida
- 01 Dióxido de Azufre (SO2) μg/m 3 Fluorescencia ultravioleta
- 06 Monóxido de Carbono (CO) mg/m3 48 Absorción infrarroja
- 07 Monóxido de Nitrógeno (NO) μg/m3 08 Quimioluminiscencia
- 08 Dióxido de Nitrógeno (NO2) μg/m 3 08 Id.
- 09 Partículas < 2.5 μm PM2.5 μg/m3 47 Microbalanza
- 10 Partículas < 10 μm PM10 μg/m3 47 Id.
- 12 Óxidos de Nitrógeno NOx μg/m3 08 Quimioluminiscencia
- 14 Ozono O3μg/m 3 06 Absorción ultravioleta
- 20 Tolueno TOL μg/m3 59 Cromatografía de gases
- 30 Benceno BEN μg/m3 59 Id.
- 35 // Etilbenceno // EBE
μg/m3
59
Id.
37
Metaxileno
MXY
μg/m3
59
Id.
38
Paraxileno
PXY
μg/m3
59
Id.
39
Ortoxileno
OXY
μg/m3
59
Id.
42
Hidrocarburos totales (hexano)
TCH
mg/m3
02
Ionización de llama
43
Metano
CH4
mg/m3
02
Id.